docs/content/sips/xxx-trigger-executors.md

```diff
@@ -0,0 +1,65 @@
+title = "SIP xxx - Trigger Executors"
+template = "main"
+
+---
+
+Summary: A new architecture for Spin trigger executors.
+
+Owner: lann.martin@fermyon.com
+
+Created: April 29, 2022
+
+<!-- Updated: April 4, 2022 -->
+
+## Background
+
+Spin currently includes built-in implementations for two trigger types: HTTP, and Redis (PubSub).
+We want to allow for additional trigger types in the future, whether they are "internal" types developed and
+supported within the Spin codebase, or "external" types developed by the broader community.
+
+While it is currently possible to build external trigger types - as in the
+[TimerTrigger example](https://github.com/fermyon/spin/blob/main/examples/spin-timer/src/main.rs) - it requires replicating much of the
+functionality provided by the `spin up` command.
+
+## Proposal
+
+### Trigger types are implemented as "trigger executors"
+
+These executors are regular subcommands/binaries, but are not normally executed directly by
+users.
+
+* "Internal" triggers use `spin` subcommands, e.g. `spin trigger http`
+* "External" triggers use external binaries, e.g. `spin-trigger-timer`
+* Trigger executors share common code, including common CLI options in a `spin-trigger` crate
```

Comment:
Do we need to say anything about versioning here - if `spin up` and `spin-trigger-foo` understand different sets of CLI options?

Reply:
I was thinking that `spin up` would only accept a couple of options itself, mostly passing through arguments to the trigger without attempting to validate. That said, I had considered setting an env var which could be e.g. `SPIN_VERSION=0.3` and allow the trigger framework to bail out if necessary.

Reply:
What would this mean for someone doing `spin up --help`?

Reply:
Great question...I suppose `spin up` and the trigger could just print their options one after the other?

Reply:
You don't have a trigger at that point though...

Reply:
It could! Though it would be weird for bindles...🤔

Reply:
Could you expand on that?  Are you thinking it would look to see if there was a spin.toml in the current directory?

(This would end up requiring a custom help implementation either way, so we can do what we like...)

Reply:
Yes, basically:

1. Print `up`'s own help
2. Run the "Detect the trigger type for an application" step (below)
3. If found, `exec` the trigger w/ `--help`

Reply:
This raises another good point, that `spin up` still needs to be responsible for fetching _at least_ `spin.toml` from bindles.

Reply:
I was assuming that `spin up` would still take care of all the asset loading, but yeah, maybe some of that needs to go to the trigger because we can't pass the abstract application object to the child process, and that means each trigger needs to know how to load all possible sources.

I wonder if we are better off defining an interface between `spin up` and external triggers, similar to the abstract application object, so that external triggers don't need to carry all that logic with them (and to update with things like performance fixes).

Reply:
I think this is another area where a prototype would help.

Reply:
I added a bit more to this section.

Reply:
`spin up` could handle the core components (that is without trigger configuration), I wonder if splitting that between `spin up` itself and the trigger is worth it though.

Just something to think about, definitely not blocking this SIP, and I think a prototype would help a lot here.

Reply:
`spin up` could handle the core components (that is without trigger configuration), I wonder if splitting that between `spin up` itself and the trigger is worth it though.

Just something to think about, definitely not blocking this SIP, and I think a prototype would help a lot here.